### PR TITLE
BC-3 Migrating Project Creation to node.js

### PR DESCRIPTION
**Pull request by ChrisP19**
This PR will also merge the http and socket.io servers. They will listen on the same port. Http requests will handle pages like project creation. socket.io will handle all loading/saving. 

Loading and saving will be handled in a future PR as follows:
- Autosaving always enabled
- Sync off: all sessions load from most recently saved file, and save to the same directory. This means new sessions will start off up to date, then stop synchronizing
- Sync on: same load/save as before, but now data is synchronized across task

Some features for project creation (like multi-level categories) will be migrated in future PRs.

### CODE REVIEW

File: app/src/js/functional/bdd_types.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  /** label attributes */
+  attributes: { [key: string]: string[] }
+  /** if shape was manual */
+  manualShape: boolean
+  /** box2d label */
+  box2d: RectType | null
+  /** poly2d label */
+  poly2d: PolygonType | null
+  /** box3d label */
+  poly3d: CubeType | null
```
Comment (gycn) [inline app/src/js/functional/bdd_types.ts]: This should be box3d

Comment (gycn) [inline app/src/js/functional/bdd_types.ts]: Should we just have a single field here instead and include the type?

File: app/src/js/functional/bdd_types.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+export interface TaskStatus {
+  /** max label id in task */
+  maxLabelId: number,
+  /** max shape id in task */
+  maxShapeId: number,
+  /** highest order in task */
+  maxOrder: number
+}
```
Comment (gycn) [inline app/src/js/functional/bdd_types.ts]: Can you use the TaskStatus? from types

File: app/src/js/components/create_form.tsx [hunk 3]
```diff
@@ -130,17 +131,17 @@ export default class CreateForm extends React.Component<Props, State> {
...
                 >
                   <option/>
-                  <option value='image'>Image</option>
-                  <option value='video'>Video</option>
-                  <option value='pointcloud'>Point Cloud</option>
-                  <option value='pointcloudtracking'>
+                  <option value={ItemTypeName.IMAGE}>Image</option>
+                  <option value={ItemTypeName.VIDEO}>Video</option>
+                  <option value={ItemTypeName.POINT_CLOUD}>Point Cloud</option>
+                  <option value={ItemTypeName.POINT_CLOUD_TRACKING}>
```
Comment (gycn) [inline app/src/js/components/create_form.tsx]: Can we have only image & point cloud here, and add a checkbox for tracking?

File: app/src/js/server/create_project.ts [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+  for (const key of Object.keys(project.items)) {
+    items = items.concat(project.items[key])
+  }
+  for (let i = 0; i < items.length; i += project.options.config.taskSize) {
+    const itemsForTask = items.slice(i, i + project.options.config.taskSize)
+    let itemInd = 0
+    for (const item of itemsForTask) {
+      item.index = itemInd
+      itemInd += 1
+    }
```
Comment (gycn) [inline app/src/js/server/create_project.ts]: Make this into a normal for (let j=0...) loop

File: app/src/js/server/types.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  'rider',
+  'car',
+  'truck',
+  'bus',
+  'train',
+  'motor',
+  'bike',
+  'traffic sign',
+  'traffic light'
+]
```
Comment (gycn) [inline app/src/js/server/types.ts]: Can you also make this the default categories for box3d?

File: app/src/js/server/main.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  path.join(__dirname, '../control'), { extensions: ['html'] }))
+app.use(express.static(
+  path.join(__dirname, '../annotation'), { extensions: ['html'] }))
+app.use('/js', express.static(path.join(__dirname, '/')))
+
+// Dynamics Handlers
+// TODO: set up other endpoint
+// TODO: cleaner way to remove leading dot from endpoint name
+app.get(Endpoint.GET_PROJECT_NAMES.substr(1),listeners.ProjectNameHandler)
+app.post(Endpoint.POST_PROJECT.substr(1), listeners.PostProjectHandler)
```
[the remark this answers is not in the document]
Comment (ChrisP19) [reply app/src/js/server/main.ts]: Fixed, turns out you never need the dot.

File: app/src/js/functional/bdd_types.ts (hunk 2 above)
Comment (ChrisP19) [reply app/src/js/functional/bdd_types.ts]: Should be able to, removed this

File: app/src/js/server/types.ts (hunk 5 above)
Comment (ChrisP19) [reply app/src/js/server/types.ts]: Renamed to defaultBoxCategories and used for both

File: app/src/js/functional/bdd_types.ts (hunk 1 above)
Comment (ChrisP19) [reply app/src/js/functional/bdd_types.ts]: @noahjimenez5844 can I change this or would that break the export type?

File: app/src/js/components/create_form.tsx (hunk 3 above)
Comment (ChrisP19) [reply app/src/js/components/create_form.tsx]: So image + tracking would be video then?

File: app/src/js/functional/bdd_types.ts (hunk 1 above)
Comment (noahjimenez5844) [reply app/src/js/functional/bdd_types.ts]: We need this as the bdd-format is split into box2d, poly2d, and poly3d: 
https://github.com/ucbdrive/bdd-data/blob/master/doc/format.md
This is the type that actually gets exported

Comment (ChrisP19) [reply app/src/js/functional/bdd_types.ts]: Apparently this comes from the bdd-format so it's necessary to have three separate fields.

Comment (noahjimenez5844) [reply app/src/js/functional/bdd_types.ts]: should be box3d but I can change this in the export PR

File: app/src/js/components/create_form.tsx (hunk 3 above)
Comment (ChrisP19) [reply app/src/js/components/create_form.tsx]: Also, I'm not sure where to put a checkbox next to item type without making the UI look pretty bad:
Screen Shot 2019-10-17 at 9 45 35 PM

File: app/src/js/functional/bdd_types.ts (hunk 1 above)
Comment (ChrisP19) [reply app/src/js/functional/bdd_types.ts]: I changed the box3d bit already

File: app/src/js/components/create_form.tsx (hunk 3 above)
Comment (noahjimenez5844) [reply app/src/js/components/create_form.tsx]: Also adding to the form could break the go project creation, so I would hesitate changing the form until we're 100% ready to switch

Comment (gycn) [reply app/src/js/components/create_form.tsx]: Sure, we can wait on adding the button, but we transform the values in the backend to image or point cloud for video and point cloud tracking, and set the tracking flag to true

Comment (ChrisP19) [reply app/src/js/components/create_form.tsx]: I see, fixed this by transforming the values in the backend.

**Pull request merged**